File: iOS/Mobile/AudioViewController.m
//
//  AudioViewController.m
//  Mobile
//
//  Created by jkh on 7/2/13.
//  Copyright (c) 2013 Ellucian. All rights reserved.
//

#import "AudioViewController.h"
#import "UIColor+SchoolCustomization.h"
#import "UIViewController+GoogleAnalyticsTrackerSupport.h"

@interface AudioViewController ()

@property (nonatomic, strong) NSTimer *sliderTimer;

@end

@implementation AudioViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.navigationController.navigationBar.translucent = NO;
    
    NSString *urlString = [self.module propertyForKey:@"audio"];
    NSURL *url = [[NSURL alloc] initWithString:urlString];
    AVURLAsset *asset = [[AVURLAsset alloc] initWithURL:url options:nil];
    AVPlayerItem* playerItem = [AVPlayerItem playerItemWithAsset:asset];
    
    
    NSArray *keys = @[@"duration"];
    [asset loadValuesAsynchronouslyForKeys:keys completionHandler:^() {
        
        NSError *error = nil;
        AVKeyValueStatus tracksStatus = [asset statusOfValueForKey:@"duration" error:&error];
        switch (tracksStatus) {
            case AVKeyValueStatusLoaded:
            {
                dispatch_async(dispatch_get_main_queue(), ^{
                    [self.playButton setEnabled:YES];
                    [self.forwardButton setEnabled:YES];
                    [self.backButton setEnabled:YES];
                    [self.seeker setEnabled:YES];
                    [self setSlider];
                });

                break;
            }
            case AVKeyValueStatusFailed:
                break;
            case AVKeyValueStatusCancelled:
                break;
            case AVKeyValueStatusUnknown:
                break;
            case AVKeyValueStatusLoading:
                break;
        }
    }];
    
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(itemDidFinishPlaying) name:AVPlayerItemDidPlayToEndTimeNotification object:playerItem];
    self.audioPlayer = [[AVPlayer alloc] initWithPlayerItem:playerItem];

    NSString *description = [self.module propertyForKey:@"description"];
    self.textLabel.text = description;
    self.textTextView.text = description;
    
    //iOS 7 needs this set again
    self.textTextView.textColor = [UIColor whiteColor];
    self.textTextView.backgroundColor = [UIColor colorWithRed:0.0f green:0.0f blue:0.0f alpha:1.0f];
    self.textTextView.font = [UIFont systemFontOfSize:17.0];

    NSString *imageUrl = [self.module propertyForKey:@"image"];
    [self.imageView loadImageFromURLString:imageUrl];
    self.title = self.module.name;
    self.seeker.thumbTintColor = [UIColor primaryColor];
    self.seeker.minimumTrackTintColor = [UIColor primaryColor];
    
    if([self.textLabel.text length] > 0) {
        UITapGestureRecognizer *tapRecognizer = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(expandText:)];
        tapRecognizer.numberOfTapsRequired = 1;
        [self.textLabelBackgroundView addGestureRecognizer:tapRecognizer];
        
        UITapGestureRecognizer *tapTextViewRecognizer = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(expandText:)];
        tapTextViewRecognizer.numberOfTapsRequired = 1;
        [self.textTextView addGestureRecognizer:tapTextViewRecognizer];
    } else {
        [self.textLabelBackgroundView setHidden:YES];
        [self.textLabel setHidden:YES];        
    }
}

-(void) viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    [self sendView:@"Audio" forModuleNamed:self.module.name];
}

- (IBAction)goBack:(id)sender {
    CMTime newTime = CMTimeMakeWithSeconds(0, 1);
    [self.audioPlayer seekToTime:newTime];
    [self updateSlider];
}

- (IBAction)goForward:(id)sender {
    CMTime newTime = CMTimeMakeWithSeconds([self durationInSeconds]+1, 1);
    [self.audioPlayer seekToTime:newTime];
    [self updateSlider];
}

- (IBAction)play:(id)sender {
    if([self isPlaying]) {
        [self.audioPlayer pause];
        [self.playButton setImage:[UIImage imageNamed:@"media_play"] forState:UIControlStateNormal];
    } else {
        [self sendEventToTracker1WithCategory:kAnalyticsCategoryUI_Action withAction:kAnalyticsActionButton_Press withLabel:@"Play button pressed" withValue:nil forModuleNamed:self.module.name];
        [self.audioPlayer play ];
        [self.playButton setImage:[UIImage imageNamed:@"media_pause"] forState:UIControlStateNormal];
    }
}

-(BOOL)isPlaying
{
    if (self.audioPlayer.currentItem && self.audioPlayer.rate != 0)
    {
        return YES;
    }
    return NO;
}

- (void)viewWillDisappear:(BOOL)animated {
    [self.sliderTimer invalidate];
    self.sliderTimer = nil;
    [self.audioPlayer pause];
    [self setImageView:nil];
    [self setSeeker:nil];

    [super viewWillDisappear:animated];
}

-(IBAction)sliding:(id)sender{
    
    CMTime newTime = CMTimeMakeWithSeconds(self.seeker.value, 1);
    [self.audioPlayer seekToTime:newTime];
}

-(void)setSlider{
    
    self.sliderTimer = [NSTimer scheduledTimerWithTimeInterval:0.1 target:self selector:@selector(updateSlider) userInfo:nil repeats:YES];
    Float64 duration = [self durationInSeconds];
    if(!isnan(duration)) {
        self.seeker.maximumValue = duration;
    } else {
        self.seeker.maximumValue = 0.0;
    }
    [self.seeker addTarget:self action:@selector(sliding:) forControlEvents:UIControlEventValueChanged];
    self.seeker.minimumValue = 0.0;
    self.seeker.continuous = YES;
}

- (void)updateSlider
{
    if([self durationInSeconds] > 0) {
        self.seeker.maximumValue = [self durationInSeconds];
        self.seeker.value = [self currentTimeInSeconds];
    } else {
        self.seeker.enabled = NO;
    }
}

- (Float64)durationInSeconds {
    Float64 dur = CMTimeGetSeconds([self.audioPlayer.currentItem duration]);
    return dur;
}


- (Float64)currentTimeInSeconds {
    Float64 dur = CMTimeGetSeconds([self.audioPlayer currentTime]);
    return dur;
}

-(void)itemDidFinishPlaying {
    [self.playButton setImage:[UIImage imageNamed:@"media_play"] forState:UIControlStateNormal];
    CMTime newTime = CMTimeMakeWithSeconds(0, 1);
    [self.audioPlayer seekToTime:newTime];
    [self updateSlider];
}

- (IBAction)expandText:(id)sender {
    
    if(self.textLabelBackgroundView.hidden) { //shrink
        self.textLabel.hidden = NO;
        self.textLabelBackgroundView.hidden = NO;
        self.textTextView.hidden = YES;
    } else { //grow
        self.textLabel.hidden = YES;
        self.textLabelBackgroundView.hidden = YES;
        self.textTextView.hidden = NO;
    }
}

@end

File: iOS/Mobile/AudioViewController.swift
//
//  AudioViewController.swift
//  Mobile
//
//  Created by Jason Hocker on 8/21/15.
//  Copyright © 2015 Ellucian Company L.P. and its affiliates. All rights reserved.
//

import Foundation
import AVFoundation
import MediaPlayer

class AudioViewController: UIViewController, AVAudioPlayerDelegate {
    
    var module : Module?
    
    var audioPlayer : AVPlayer?
    var sliderTimer : NSTimer?
    var mpArtwork : MPMediaItemArtwork?
    
    @IBOutlet var imageView: UIImageView!
    
    @IBOutlet var seeker: UISlider!
    @IBOutlet var playButton: UIButton!
    @IBOutlet var backButton: UIButton!
    @IBOutlet var forwardButton: UIButton!
    @IBOutlet var textTextView: UITextView!
    
    @IBOutlet var textLabelBackgroundView: UIView!
    @IBOutlet var textLabel: UILabel!
    @IBOutlet var play: UIButton!
    
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        sendView("Audio", forModuleNamed: module!.name)
    }
    
    override func viewDidAppear(animated: Bool) {
        super.viewDidAppear(animated)
        UIApplication.sharedApplication().beginReceivingRemoteControlEvents()
        self.becomeFirstResponder()
    }
    
    override func viewWillDisappear(animated: Bool) {
        super.viewWillDisappear(animated)
        UIApplication.sharedApplication().endReceivingRemoteControlEvents()
        self.resignFirstResponder()
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        let urlString = module!.propertyForKey("audio")
        let url = NSURL(string: urlString)
        if let url = url {
            let asset = AVURLAsset(URL: url)
            let playerItem = AVPlayerItem(asset: asset)
            
            do {
                NSNotificationCenter.defaultCenter().addObserver(self, selector:Selector("itemDidFinishPlaying:"), name:AVPlayerItemDidPlayToEndTimeNotification, object: playerItem)
                
                sliderTimer = NSTimer .scheduledTimerWithTimeInterval(0.1, target: self, selector: Selector("updateSlider"), userInfo: nil, repeats: true)
                
                try AVAudioSession.sharedInstance().setCategory(AVAudioSessionCategoryPlayback)
                try AVAudioSession.sharedInstance().setActive(true)
                
                audioPlayer = AVPlayer(playerItem: playerItem)
                audioPlayer?.currentItem?.addObserver(self, forKeyPath: "status", options: ([.New, .Initial]), context: nil)
                
               
                if let description = module!.propertyForKey("description") where description.characters.count > 0 {
                    
                    textLabel.text = description
                    textTextView.text = description
                    
                    textTextView.textColor = UIColor.whiteColor()
                    textTextView.font = UIFont.preferredFontForTextStyle(UIFontTextStyleBody)
                    
                    let labelTapGestureRecognizer = UITapGestureRecognizer(target: self, action: Selector("expandText:"))
                    labelTapGestureRecognizer.numberOfTapsRequired = 1
                    textLabelBackgroundView.addGestureRecognizer(labelTapGestureRecognizer)
                    
                    let textViewTapGestureRecognizer = UITapGestureRecognizer(target:self, action: Selector("expandText:"))
                    textViewTapGestureRecognizer.numberOfTapsRequired = 1
                    textTextView.addGestureRecognizer(textViewTapGestureRecognizer)
                    
                    
                } else {
                    textLabelBackgroundView.hidden = true
                    textLabel.hidden = true
                }

                if let imageUrl = self.module?.propertyForKey("image") {
                    NSURLSession.sharedSession().dataTaskWithURL(NSURL(string: imageUrl)!) { (data, response, error) in
                        if let data = data {
                            dispatch_async(dispatch_get_main_queue()) {
                                let image = UIImage(data: data)
                                if let image = image {
                                    self.imageView.image = image
                                    self.mpArtwork = MPMediaItemArtwork(image: image)
                                    self.updateNowPlaying()
                                }
                            }
                        }
                    }.resume()
                }
                
                self.title = module!.name
                self.seeker.thumbTintColor = UIColor.primaryColor()
                self.seeker.minimumTrackTintColor = UIColor.primaryColor()

                self.updateNowPlaying()
            } catch let error {
                print (error)
            }
            
        }
    }
    
    // MPMediaItemPropertyAlbumTitle
    // MPMediaItemPropertyAlbumTrackCount
    // MPMediaItemPropertyAlbumTrackNumber
    // MPMediaItemPropertyArtist
    // MPMediaItemPropertyArtwork
    // MPMediaItemPropertyComposer
    // MPMediaItemPropertyDiscCount
    // MPMediaItemPropertyDiscNumber
    // MPMediaItemPropertyGenre
    // MPMediaItemPropertyPersistentID
    // MPMediaItemPropertyPlaybackDuration
    // MPMediaItemPropertyTitle
    func updateNowPlaying() {
        
        let playbackDuration = CMTimeGetSeconds(audioPlayer!.currentItem!.duration)
        let playbackTime = CMTimeGetSeconds(audioPlayer!.currentItem!.currentTime())
        
        if let artwork = self.mpArtwork{
            MPNowPlayingInfoCenter.defaultCenter().nowPlayingInfo = [ MPMediaItemPropertyArtwork : artwork, MPMediaItemPropertyTitle : module!.name, MPMediaItemPropertyPlaybackDuration:playbackDuration, MPNowPlayingInfoPropertyPlaybackRate: 1.0, MPNowPlayingInfoPropertyElapsedPlaybackTime: playbackTime ]
        
        } else {
             MPNowPlayingInfoCenter.defaultCenter().nowPlayingInfo = [ MPMediaItemPropertyTitle : module!.name, MPMediaItemPropertyPlaybackDuration:playbackDuration, MPNowPlayingInfoPropertyPlaybackRate: 1.0, MPNowPlayingInfoPropertyElapsedPlaybackTime: playbackTime ]
        }
    }
    

    
    @IBAction func togglePlay(sender: AnyObject) {
        if let _ = audioPlayer {
            if isPlaying() {
                goPause(sender)
            } else {
                goPlay(sender)
            }
        }
    }
    
    func goPlay(sender: AnyObject) {
        if let player = audioPlayer {
            sendEventToTracker1WithCategory(kAnalyticsCategoryUI_Action, withAction: kAnalyticsActionButton_Press, withLabel: "Play button pressed", withValue: nil, forModuleNamed: self.module!.name)
            player.play()
            playButton.setImage(UIImage(named: "media_pause"), forState: .Normal)
            updateNowPlaying()
        }
    }
    
    func goPause(sender: AnyObject) {
        if let player = audioPlayer {
            player.pause()
            playButton.setImage(UIImage(named: "media_play"), forState: .Normal)
            updateNowPlaying()
        }
    }
    
    @IBAction func goBack(sender: AnyObject) {
        if let player = audioPlayer {
            let newTime = CMTimeMakeWithSeconds(0, 1)
            player.seekToTime(newTime)
            updateSlider()
        }
    }
    @IBAction func goForward(sender: AnyObject) {
        if let player = audioPlayer {
            let newTime = CMTimeMakeWithSeconds(durationInSeconds() + 1, 1)
            player.seekToTime(newTime)
            updateSlider()
        }
    }
    
    func isPlaying() -> Bool {
        if let player = audioPlayer {
            return player.currentItem != nil && player.rate != 0
        }
        return false
    }
    
    func updateSlider() {
        let duration = durationInSeconds()
        if duration > 0 {
            self.seeker.maximumValue = Float(duration)
            self.seeker.value = Float(currentTimeInSeconds())
            updateNowPlaying()
        } else {
            self.seeker.enabled = false
        }
    }
    
    func durationInSeconds() -> Float64 {
        if let currentItem = audioPlayer?.currentItem {
            return CMTimeGetSeconds(currentItem.duration)
        }
        return 0
    }
    
    func currentTimeInSeconds() -> Float64 {
        if let currentItem = audioPlayer?.currentItem {
            return CMTimeGetSeconds(currentItem.currentTime())
        }
        return 0
    }
    
    @IBAction func sliding(sender: AnyObject) {
        if let player = audioPlayer {
            
            let newTime = CMTimeMakeWithSeconds(Float64(seeker.value), 1)
            player.seekToTime(newTime)
        }
    }
    
    override func observeValueForKeyPath(keyPath: String?, ofObject object: AnyObject?, change: [String : AnyObject]?, context: UnsafeMutablePointer<Void>) {
        if self.audioPlayer?.currentItem!.status == .ReadyToPlay {
            let _ = self.audioPlayer?.currentItem?.duration
            
            playButton.enabled = true
            forwardButton.enabled = true
            backButton.enabled = true
            seeker.enabled = true
            
            updateSlider()
            
            if let currentItem = self.audioPlayer?.currentItem {
                currentItem.removeObserver(self, forKeyPath: "status")
            }
        } else if self.audioPlayer?.currentItem!.status == .Failed {
            let error = self.audioPlayer?.currentItem?.error?.localizedDescription
            let alertController = UIAlertController(title: NSLocalizedString("Error Loading Audio", comment: "title when error loading audio"), message: error, preferredStyle: .Alert)
            
            let OKAction = UIAlertAction(title: "OK", style: .Default, handler: nil)
            alertController.addAction(OKAction)
            
            if let currentItem = self.audioPlayer?.currentItem {
                currentItem.removeObserver(self, forKeyPath: "status")
            }
            
            self.presentViewController(alertController, animated: true, completion: nil)
        }
    }
    
    func itemDidFinishPlaying(sender: AnyObject) {
        if let player = audioPlayer {
            playButton.setImage(UIImage(named: "media_play"), forState: .Normal)
            let newTime = CMTimeMakeWithSeconds(0, 1)
            player.seekToTime(newTime)
            updateSlider()
        }
    }
    
    override func canBecomeFirstResponder() -> Bool {
        return true
    }
    
    override func remoteControlReceivedWithEvent(event: UIEvent?) {
        if let event = event {
            if event.type == UIEventType.RemoteControl {
                switch event.subtype {
                case .RemoteControlPlay:
                    goPlay(event)
                case .RemoteControlPause, .RemoteControlStop:
                    goPause(event)
                case .RemoteControlTogglePlayPause:
                   togglePlay(event)
                case .RemoteControlNextTrack:
                    goForward(event)
                case .RemoteControlPreviousTrack:
                    goBack(event)
                case .RemoteControlBeginSeekingBackward, .RemoteControlEndSeekingBackward, .RemoteControlBeginSeekingForward, .RemoteControlEndSeekingForward:
                    break
                default:
                    break
                }
            }
        }
    }
    
    func expandText(sender: AnyObject) {
        if textLabelBackgroundView.hidden { //shrink
            textLabel.hidden = false
            textLabelBackgroundView.hidden = false
            textTextView.hidden = true
        } else { //grow
            textLabel.hidden = true
            textLabelBackgroundView.hidden = true
            textTextView.hidden = false
        }
    }
}